src/awkward1/operations/structure.py

```diff
@@ -744,20 +740,49 @@ def concatenate(arrays, axis=0, mergebool=True, highlevel=True):
             + awkward1._util.exception_suffix(__file__)
         )
 
-    batch = [contents[0]]
-    for x in contents[1:]:
-        if batch[-1].mergeable(x, mergebool=mergebool):
-            batch.append(x)
-        else:
-            collapsed = batch[0].mergemany(batch[1:])
-            batch = [collapsed.merge_as_union(x)]
+    if axis < 0:
+        raise ValueError(
+            "the 'axis' of concatenate must be non-negative"
+            + awkward1._util.exception_suffix(__file__)
+        )
```

Comment:
Is this necessary? Your implementation uses `axis_wrap_if_negative`.

Reply:
Oh! Is it because you check

```python
if depth == axis
```

in `getfunction`? Maybe you could expose the `axis_wrap_if_negative` to Python, since this needs to be checked while descending, not when you've already reached the target axis. That could enable the feature for `axis < 0`.